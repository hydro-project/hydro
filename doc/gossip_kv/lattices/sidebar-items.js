window.SIDEBAR_ITEMS = {"struct":["BoundedSetLattice"]};