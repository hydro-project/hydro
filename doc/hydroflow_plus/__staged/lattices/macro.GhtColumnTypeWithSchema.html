<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=../../../hydroflow_plus/lattices/macro.GhtColumnTypeWithSchema.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="../../../hydroflow_plus/lattices/macro.GhtColumnTypeWithSchema.html">../../../hydroflow_plus/lattices/macro.GhtColumnTypeWithSchema.html</a>...</p>
    <script>location.replace("../../../hydroflow_plus/lattices/macro.GhtColumnTypeWithSchema.html" + location.search + location.hash);</script>
</body>
</html>