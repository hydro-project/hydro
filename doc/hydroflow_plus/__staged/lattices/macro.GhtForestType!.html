<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.GhtForestType.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.GhtForestType.html">macro.GhtForestType.html</a>...</p>
    <script>location.replace("macro.GhtForestType.html" + location.search + location.hash);</script>
</body>
</html>