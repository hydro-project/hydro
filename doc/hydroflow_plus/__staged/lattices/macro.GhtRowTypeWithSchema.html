<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=../../../hydroflow_plus/lattices/macro.GhtRowTypeWithSchema.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="../../../hydroflow_plus/lattices/macro.GhtRowTypeWithSchema.html">../../../hydroflow_plus/lattices/macro.GhtRowTypeWithSchema.html</a>...</p>
    <script>location.replace("../../../hydroflow_plus/lattices/macro.GhtRowTypeWithSchema.html" + location.search + location.hash);</script>
</body>
</html>