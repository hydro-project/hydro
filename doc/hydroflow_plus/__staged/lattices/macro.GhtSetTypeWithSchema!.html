<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.GhtSetTypeWithSchema.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.GhtSetTypeWithSchema.html">macro.GhtSetTypeWithSchema.html</a>...</p>
    <script>location.replace("macro.GhtSetTypeWithSchema.html" + location.search + location.hash);</script>
</body>
</html>