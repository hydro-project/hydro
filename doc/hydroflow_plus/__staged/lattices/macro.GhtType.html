<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=../../../hydroflow_plus/lattices/macro.GhtType.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="../../../hydroflow_plus/lattices/macro.GhtType.html">../../../hydroflow_plus/lattices/macro.GhtType.html</a>...</p>
    <script>location.replace("../../../hydroflow_plus/lattices/macro.GhtType.html" + location.search + location.hash);</script>
</body>
</html>