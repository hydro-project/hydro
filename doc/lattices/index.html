<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="The `lattices` crate provides ergonomic and composable lattice types. You can also implement custom lattices via a few simple traits."><title>lattices - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-e56847b5.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="lattices" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (9f32ccf35 2025-09-21)" data-channel="nightly" data-search-js="search-e256b49e.js" data-stringdex-js="stringdex-061df703.js" data-settings-js="settings-c38705f0.js" ><script src="../static.files/storage-e2aeef58.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-6dc2a7f3.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Crate lattices</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../lattices/index.html">lattices</a><span class="version">0.6.1</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section id="rustdoc-toc"><h3><a href="#">Sections</a></h3><ul class="block top-toc"><li><a href="#lattices" title="Lattices">Lattices</a></li><li><a href="#traits" title="Traits">Traits</a><ul><li><a href="#merge" title="`Merge`"><code>Merge</code></a></li><li><a href="#partialord-latticeord-and-naivelatticeord" title="`PartialOrd`, `LatticeOrd`, and `NaiveLatticeOrd`"><code>PartialOrd</code>, <code>LatticeOrd</code>, and <code>NaiveLatticeOrd</code></a></li><li><a href="#latticefrom" title="`LatticeFrom`"><code>LatticeFrom</code></a></li><li><a href="#isbot-istop-and-default" title="`IsBot`, `IsTop`, and `Default`"><code>IsBot</code>, <code>IsTop</code>, and <code>Default</code></a></li><li><a href="#atomize" title="`Atomize`"><code>Atomize</code></a></li><li><a href="#deepreveal" title="`DeepReveal`"><code>DeepReveal</code></a></li></ul></li></ul><h3><a href="#reexports">Crate Items</a></h3><ul class="block"><li><a href="#reexports" title="Re-exports">Re-exports</a></li><li><a href="#modules" title="Modules">Modules</a></li><li><a href="#macros" title="Macros">Macros</a></li><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#traits-1" title="Traits">Traits</a></li><li><a href="#functions" title="Functions">Functions</a></li><li><a href="#derives" title="Derive Macros">Derive Macros</a></li></ul></section><div id="rustdoc-modnav"></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><h1>Crate <span>lattices</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/lattices/lib.rs.html#1-269">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><h1 class="crate-title">The <code>lattices</code> Crate</h1>
<p>The <code>lattices</code> crate provides ergonomic and composable lattice types. You can also implement custom
lattices via a few simple traits.</p>
<p>Lattices are an incredibly powerful mathematical concept which can greatly simplify the trickiness
of distributed computing. They align very well with the reality of what happens physically in a
distributed system: messages can always arrive out-of-order or duplicated. But if that data is
represented as lattices then all machines will always reach the same end result simply by merging the data together.
One popular way that lattices are currently used in distributed systems is as the data underlying
<a href="https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type">Conflict-free Replicated Data Types</a>
(CRDTs).</p>
<p>Lattices also allow us to harness the power of the CALM Theorem: “a program has a consistent,
coordination-free distributed implementation if and only if it is monotonic.” Lattice state is
always monotonic, meaning any part of a distributed system built on lattice state can be
freely distributed with no coordination overhead. The goal of the <a href="https://hydro.run/">Hydro Project</a>
is to allow users to write programs that automatically scale and distribute effortlessly.</p>
<p>For more information on the underlying mathematics of lattices and monotonicity, take a look at
<a href="https://hydro.run/docs/dfir/lattices_crate/lattice_math">Lattice Math section of the Hydroflow Book</a>
and Section 2 of the <a href="https://hydro.run/papers/hydroflow-thesis.pdf">Hydroflow Thesis (2021)</a>.</p>
<p>Take a look at the <a href="https://hydro.run/rustdoc/lattices/index.html"><code>lattice</code> rustdocs</a>.</p>
<h3 id="lattices"><a class="doc-anchor" href="#lattices">§</a>Lattices</h3>
<p><code>lattices</code> provides implementations of common lattice types:</p>
<ul>
<li><a href="struct.Min.html" title="struct lattices::Min"><code>Min&lt;T&gt;</code></a> and <a href="struct.Max.html" title="struct lattices::Max"><code>Max&lt;T&gt;</code></a> - totally-ordered lattices.</li>
<li><a href="set_union/struct.SetUnion.html" title="struct lattices::set_union::SetUnion"><code>set_union::SetUnion&lt;T&gt;</code></a> - set-union lattice of scalar values.</li>
<li><a href="map_union/struct.MapUnion.html" title="struct lattices::map_union::MapUnion"><code>map_union::MapUnion&lt;K, Lat&gt;</code></a> - scalar keys with nested lattice values.</li>
<li><a href="union_find/struct.UnionFind.html" title="struct lattices::union_find::UnionFind"><code>union_find::UnionFind&lt;K&gt;</code></a> - union partitions of a set of scalar values.</li>
<li><a href="struct.VecUnion.html" title="struct lattices::VecUnion"><code>VecUnion&lt;Lat&gt;</code></a> - growing <code>Vec</code> of nested lattices, like <code>MapUnion&lt;&lt;usize, Lat&gt;&gt;</code> but without missing entries.</li>
<li><a href="struct.WithBot.html" title="struct lattices::WithBot"><code>WithBot&lt;Lat&gt;</code></a> - wraps a lattice in <code>Option</code> with <code>None</code> as the new bottom value.</li>
<li><a href="struct.WithTop.html" title="struct lattices::WithTop"><code>WithTop&lt;Lat&gt;</code></a> - wraps a lattice in <code>Option</code> with <code>None</code> as the new <em>top</em> value.</li>
<li><a href="struct.Pair.html" title="struct lattices::Pair"><code>Pair&lt;LatA, LatB&gt;</code></a> - product of two nested lattices.</li>
<li><a href="struct.DomPair.html" title="struct lattices::DomPair"><code>DomPair&lt;LatKey, LatVal&gt;</code></a>* - a versioned pair where the <code>LatKey</code> dominates the <code>LatVal</code>.</li>
<li><a href="struct.Conflict.html" title="struct lattices::Conflict"><code>Conflict&lt;T&gt;</code></a>* - adds a “conflict” top to domain <code>T</code>. Merging inequal <code>T</code>s results in top.</li>
<li><a href="struct.Point.html" title="struct lattices::Point"><code>Point&lt;T, *&gt;</code></a>* - a single “point lattice” value which cannot be merged with any inequal value.</li>
<li><a href="https://doc.rust-lang.org/std/primitive.unit.html"><code>()</code></a> - the “unit” lattice, a “point lattice” with unit <code>()</code> as the only value in the domain.</li>
</ul>
<p>*Special implementations which do not obey all lattice properties but are still useful under
certain circumstances.</p>
<p>Additionally, custom lattices can be made by implementing the traits below.</p>
<h3 id="traits"><a class="doc-anchor" href="#traits">§</a>Traits</h3>
<p>A type becomes a lattice by implementing one or more traits starting with <code>Merge</code>. These traits
are already implemented for all the provided lattice types.</p>
<h4 id="merge"><a class="doc-anchor" href="#merge">§</a><code>Merge</code></h4>
<p>The main trait is <a href="trait.Merge.html" title="trait lattices::Merge"><code>Merge</code></a>, which defines a lattice merge function (AKA “join” or “least upper
bound”). Implementors must define the <a href="trait.Merge.html#tymethod.merge" title="method lattices::Merge::merge"><code>Merge::merge</code></a> method which does a merge in-place into
<code>&amp;mut self</code>. The method must return <code>true</code> if <code>self</code> was modified (i.e. the value got larger in the
lattice partial order) and <code>false</code> otherwise (i.e. <code>other</code> was smaller than <code>self</code>). The <a href="trait.Merge.html#method.merge_owned" title="associated function lattices::Merge::merge_owned"><code>Merge::merge_owned</code></a>
function, which merges two owned values, is provided.</p>
<p>The <code>merge</code> method must be associative, commutative, and idempotent. This is not checked by the
compiler, but the implementor can use the <a href="test/fn.check_lattice_properties.html" title="fn lattices::test::check_lattice_properties"><code>test::check_lattice_properties</code></a> method to spot-check
these properties on a collection of values.</p>
<h4 id="partialord-latticeord-and-naivelatticeord"><a class="doc-anchor" href="#partialord-latticeord-and-naivelatticeord">§</a><code>PartialOrd</code>, <code>LatticeOrd</code>, and <code>NaiveLatticeOrd</code></h4>
<p>Rust already has a trait for partial orders, <a href="https://doc.rust-lang.org/nightly/core/cmp/trait.PartialOrd.html" title="trait core::cmp::PartialOrd"><code>PartialOrd</code></a>, which should be implemented on lattice
types. However that trait is not specific to lattice partial orders, therefore we provide the<a href="trait.LatticeOrd.html" title="trait lattices::LatticeOrd"><code>LatticeOrd&lt;Rhs&gt;</code></a><code>: PartialOrd&lt;Rhs&gt;</code>
marker trait to denote when a <code>PartialOrd</code> implementation is a lattice partial order. <code>LatticeOrd</code>
must always agree with the <code>Merge</code> function.</p>
<p>Additionally, the sealed <a href="trait.NaiveLatticeOrd.html" title="trait lattices::NaiveLatticeOrd"><code>NaiveLatticeOrd</code></a> trait is provided on all lattice types that implement
<code>Merge</code> and <code>Clone</code>. This trait provides a <code>naive_cmp</code> method which derives a lattice order from
the <code>Merge</code> function directly. However the implementation is generally slow and inefficient.</p>
<p>Implementors should use the <a href="test/fn.check_partial_ord_properties.html" title="fn lattices::test::check_partial_ord_properties"><code>test::check_partial_ord_properties</code></a> method to check their
<code>PartialOrd</code> implementation, and should use the <a href="test/fn.check_lattice_ord.html" title="fn lattices::test::check_lattice_ord"><code>test::check_lattice_ord</code></a> to ensure the partial
order agrees with the <code>Merge</code>-derived <code>NaiveLatticeOrd</code> order.</p>
<h4 id="latticefrom"><a class="doc-anchor" href="#latticefrom">§</a><code>LatticeFrom</code></h4>
<p><a href="trait.LatticeFrom.html" title="trait lattices::LatticeFrom"><code>LatticeFrom</code></a> is equivalent to the <a href="https://doc.rust-lang.org/nightly/core/convert/trait.From.html" title="trait core::convert::From"><code>std::convert::From</code></a> trait but specific to lattices.
<code>LatticeFrom</code> should be implemented only between different representations of the same lattice
type, e.g. between <a href="set_union/type.SetUnionBTreeSet.html" title="type lattices::set_union::SetUnionBTreeSet"><code>set_union::SetUnionBTreeSet</code></a> and <a href="set_union/type.SetUnionHashSet.html" title="type lattices::set_union::SetUnionHashSet"><code>set_union::SetUnionHashSet</code></a>. For compound
lattice (lattices with nested lattice types), the <code>LatticeFrom</code> implementation should be recursive
for those nested lattices.</p>
<h4 id="isbot-istop-and-default"><a class="doc-anchor" href="#isbot-istop-and-default">§</a><code>IsBot</code>, <code>IsTop</code>, and <code>Default</code></h4>
<p>A bottom (⊥) is strictly less than all other values. A top (⊤) is strictly greater than all other
values. <code>IsBot::is_bot</code> and <code>IsTop::is_top</code> determine if a lattice instance is top or bottom
respectively.</p>
<p>For lattice types, <code>Default::default()</code> must create a bottom value. <code>IsBot::is_bot(&amp;Default::default())</code>
should always return true for all lattice types.</p>
<h4 id="atomize"><a class="doc-anchor" href="#atomize">§</a><code>Atomize</code></h4>
<p><a href="trait.Atomize.html#tymethod.atomize" title="method lattices::Atomize::atomize"><code>Atomize::atomize</code></a> converts a lattice point into a bunch of smaller lattice points. When these
“atoms” are merged together they will form the original lattice point. See the docs for more
precise semantics.</p>
<h4 id="deepreveal"><a class="doc-anchor" href="#deepreveal">§</a><code>DeepReveal</code></h4>
<p><a href="trait.DeepReveal.html" title="trait lattices::DeepReveal"><code>DeepReveal</code></a> allows recursive “revealing” of the underlying data within latties. Particularly
useful for revealing nested lattices.</p>
</div></details><h2 id="reexports" class="section-header">Re-exports<a href="#reexports" class="anchor">§</a></h2><dl class="item-table reexports"><dt id="reexport.cc_traits"><code>pub use cc_traits;</code></dt><dt id="reexport.variadics"><code>pub use variadics;</code></dt></dl><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">§</a></h2><dl class="item-table"><dt><a class="mod" href="algebra/index.html" title="mod lattices::algebra">algebra</a></dt><dd>Module for definiting algebraic structures and properties.</dd><dt><a class="mod" href="collections/index.html" title="mod lattices::collections">collections</a></dt><dd>Simple singleton or array collection with [<code>cc_traits</code>] implementations.</dd><dt><a class="mod" href="ght/index.html" title="mod lattices::ght">ght</a></dt><dd>GHT from the Wang/Willsey/Suciu Freejoin work</dd><dt><a class="mod" href="map_union/index.html" title="mod lattices::map_union">map_<wbr>union</a></dt><dd>Module containing the <a href="map_union/struct.MapUnion.html" title="struct lattices::map_union::MapUnion"><code>MapUnion</code></a> lattice and aliases for different datastructures.</dd><dt><a class="mod" href="map_union_with_tombstones/index.html" title="mod lattices::map_union_with_tombstones">map_<wbr>union_<wbr>with_<wbr>tombstones</a></dt><dd>Module containing the <a href="map_union_with_tombstones/struct.MapUnionWithTombstones.html" title="struct lattices::map_union_with_tombstones::MapUnionWithTombstones"><code>MapUnionWithTombstones</code></a> lattice and aliases for different datastructures.</dd><dt><a class="mod" href="semiring_application/index.html" title="mod lattices::semiring_application">semiring_<wbr>application</a></dt><dd>Module containing the <a href="semiring_application/struct.BinaryTrust.html" title="struct lattices::semiring_application::BinaryTrust"><code>BinaryTrust</code></a> applications.</dd><dt><a class="mod" href="set_union/index.html" title="mod lattices::set_union">set_<wbr>union</a></dt><dd>Module containing the <a href="set_union/struct.SetUnion.html" title="struct lattices::set_union::SetUnion"><code>SetUnion</code></a> lattice and aliases for different datastructures.</dd><dt><a class="mod" href="set_union_with_tombstones/index.html" title="mod lattices::set_union_with_tombstones">set_<wbr>union_<wbr>with_<wbr>tombstones</a></dt><dd>Module containing the <a href="set_union_with_tombstones/struct.SetUnionWithTombstones.html" title="struct lattices::set_union_with_tombstones::SetUnionWithTombstones"><code>SetUnionWithTombstones</code></a> lattice and aliases for different datastructures.</dd><dt><a class="mod" href="test/index.html" title="mod lattices::test">test</a></dt><dd>Helper test utils to test lattice implementation correctness.</dd><dt><a class="mod" href="union_find/index.html" title="mod lattices::union_find">union_<wbr>find</a></dt><dd>Module containing the <a href="union_find/struct.UnionFind.html" title="struct lattices::union_find::UnionFind"><code>UnionFind</code></a> lattice and aliases for different datastructures.</dd></dl><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><dl class="item-table"><dt><a class="macro" href="macro.ColtType.html" title="macro lattices::ColtType">Colt<wbr>Type</a></dt><dd>Construct a forest of Ghts (i.e. a ColtForest) with the given schema and storage type.</dd><dt><a class="macro" href="macro.GhtType.html" title="macro lattices::GhtType">GhtType</a></dt><dd>Public macro for constructing a Ght struct with the given schema and storage type</dd><dt><a class="macro" href="macro.GhtTypeWithSchema.html" title="macro lattices::GhtTypeWithSchema">GhtType<wbr>With<wbr>Schema</a></dt><dd>Internal macro for constructing a Ght struct with the given schema and storage type</dd></dl><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><dl class="item-table"><dt><a class="struct" href="struct.Conflict.html" title="struct lattices::Conflict">Conflict</a></dt><dd>A <code>Conflict</code> lattice, stores a single instance of <code>T</code> and goes to a “conflict” state (<code>None</code>)
if inequal <code>T</code> instances are merged together.</dd><dt><a class="struct" href="struct.DomPair.html" title="struct lattices::DomPair">DomPair</a></dt><dd>Dominating pair compound lattice.</dd><dt><a class="struct" href="struct.Max.html" title="struct lattices::Max">Max</a></dt><dd>A totally ordered max lattice. Merging returns the larger value.</dd><dt><a class="struct" href="struct.Min.html" title="struct lattices::Min">Min</a></dt><dd>A totally ordered min lattice. Merging returns the smaller value.</dd><dt><a class="struct" href="struct.Pair.html" title="struct lattices::Pair">Pair</a></dt><dd>Pair compound lattice.</dd><dt><a class="struct" href="struct.PairBimorphism.html" title="struct lattices::PairBimorphism">Pair<wbr>Bimorphism</a></dt><dd>Bimorphism which pairs up the two input lattices.</dd><dt><a class="struct" href="struct.Point.html" title="struct lattices::Point">Point</a></dt><dd>A <code>Point</code> lattice, corresponding to a single instance of <code>T</code>.</dd><dt><a class="struct" href="struct.VecUnion.html" title="struct lattices::VecUnion">VecUnion</a></dt><dd>Vec-union compound lattice.</dd><dt><a class="struct" href="struct.WithBot.html" title="struct lattices::WithBot">WithBot</a></dt><dd>Adds a new “bot” value to the nested lattice type.</dd><dt><a class="struct" href="struct.WithTop.html" title="struct lattices::WithTop">WithTop</a></dt><dd>Adds a new “top” value to the nested lattice type.</dd></dl><h2 id="traits-1" class="section-header">Traits<a href="#traits-1" class="anchor">§</a></h2><dl class="item-table"><dt><a class="trait" href="trait.Addition.html" title="trait lattices::Addition">Addition</a></dt><dd>Trait for Semiring Addition.</dd><dt><a class="trait" href="trait.Atomize.html" title="trait lattices::Atomize">Atomize</a></dt><dd>Trait to atomize a lattice into individual elements. For example, a <a href="set_union/struct.SetUnion.html" title="struct lattices::set_union::SetUnion"><code>set_union::SetUnion</code></a>
will be broken up into individual singleton elements.</dd><dt><a class="trait" href="trait.DeepReveal.html" title="trait lattices::DeepReveal">Deep<wbr>Reveal</a></dt><dd>Trait for recursively revealing the underlying types within lattice types.</dd><dt><a class="trait" href="trait.IsBot.html" title="trait lattices::IsBot">IsBot</a></dt><dd>Trait to check if a lattice instance is bottom (⊥).</dd><dt><a class="trait" href="trait.IsTop.html" title="trait lattices::IsTop">IsTop</a></dt><dd>Trait to check if a lattice instance is top (⊤) and therefore cannot change any futher.</dd><dt><a class="trait" href="trait.Lattice.html" title="trait lattices::Lattice">Lattice</a></dt><dd>Alias trait for lattice types.</dd><dt><a class="trait" href="trait.LatticeBimorphism.html" title="trait lattices::LatticeBimorphism">Lattice<wbr>Bimorphism</a></dt><dd>Semilattice bimorphism. Lattice merge must distribute over this binary function, in both arguments.</dd><dt><a class="trait" href="trait.LatticeFrom.html" title="trait lattices::LatticeFrom">Lattice<wbr>From</a></dt><dd>Same as <code>From</code> but for lattices.</dd><dt><a class="trait" href="trait.LatticeMorphism.html" title="trait lattices::LatticeMorphism">Lattice<wbr>Morphism</a></dt><dd>Semilattice morphism. Lattice merge must distribute over this unary function.</dd><dt><a class="trait" href="trait.LatticeOrd.html" title="trait lattices::LatticeOrd">Lattice<wbr>Ord</a></dt><dd>Trait for lattice partial order comparison
PartialOrd is implemented for many things, this trait can be used to require the type be a lattice.</dd><dt><a class="trait" href="trait.Merge.html" title="trait lattices::Merge">Merge</a></dt><dd>Trait for lattice merge (AKA “join” or “least upper bound”).</dd><dt><a class="trait" href="trait.Multiplication.html" title="trait lattices::Multiplication">Multiplication</a></dt><dd>Trait for Semiring Multiplication.</dd><dt><a class="trait" href="trait.NaiveLatticeOrd.html" title="trait lattices::NaiveLatticeOrd">Naive<wbr>Lattice<wbr>Ord</a></dt><dd>Naive lattice compare, based on the <a href="trait.Merge.html#tymethod.merge" title="method lattices::Merge::merge"><code>Merge::merge</code></a> function.</dd><dt><a class="trait" href="trait.One.html" title="trait lattices::One">One</a></dt><dd>Trait to define a one in a semiring.</dd><dt><a class="trait" href="trait.Semiring.html" title="trait lattices::Semiring">Semiring</a></dt><dd>Alias trait for semirings.</dd><dt><a class="trait" href="trait.Zero.html" title="trait lattices::Zero">Zero</a></dt><dd>Trait to check if semiring contains a zero.</dd></dl><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><dl class="item-table"><dt><a class="fn" href="fn.closure_to_bimorphism.html" title="fn lattices::closure_to_bimorphism">closure_<wbr>to_<wbr>bimorphism</a></dt><dd>Converts a closure to a bimorphism. Does not check for correctness.</dd><dt><a class="fn" href="fn.closure_to_morphism.html" title="fn lattices::closure_to_morphism">closure_<wbr>to_<wbr>morphism</a></dt><dd>Converts a closure to a morphism. Does not check for correctness.</dd></dl><h2 id="derives" class="section-header">Derive Macros<a href="#derives" class="anchor">§</a></h2><dl class="item-table"><dt><a class="derive" href="derive.IsBot.html" title="derive lattices::IsBot">IsBot</a></dt><dd>Derives lattice <code>IsBot</code>.</dd><dt><a class="derive" href="derive.IsTop.html" title="derive lattices::IsTop">IsTop</a></dt><dd>Derives lattice <code>IsTop</code>.</dd><dt><a class="derive" href="derive.Lattice.html" title="derive lattices::Lattice">Lattice</a></dt><dd><code>#[derive(Lattice)]</code> Macro</dd><dt><a class="derive" href="derive.LatticeFrom.html" title="derive lattices::LatticeFrom">Lattice<wbr>From</a></dt><dd>Derives <code>LatticeFrom</code>.</dd><dt><a class="derive" href="derive.LatticeOrd.html" title="derive lattices::LatticeOrd">Lattice<wbr>Ord</a></dt><dd>Derives [<code>PartialEq</code>], [<code>PartialOrd</code>], and <code>LatticeOrd</code> together.</dd><dt><a class="derive" href="derive.Merge.html" title="derive lattices::Merge">Merge</a></dt><dd>Derives lattice <code>Merge</code>.</dd></dl></section></div></main></body></html>