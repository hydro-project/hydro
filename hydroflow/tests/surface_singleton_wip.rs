use lattices::Max;
use multiplatform_test::multiplatform_test;

#[multiplatform_test]
pub fn test_fold_tick() {
    let mut df = hydroflow::hydroflow_syntax! {
        stream1 = source_iter(10..=30);
        stream2 = source_iter_delta(15..=25) -> map(Max::new);
        sum_of_stream2 = stream2 -> state::<Max<_>>();

        sum_of_stream2[items] -> null();
        sum_of_stream2_reference = sum_of_stream2[state];
        sum_of_stream2_reference -> null::<()>();

        // Do we need to stratify between sum_of_stream2_reference and it's usage in the filter?
        // Mark state as blocking and just move on.
        // (Little different - output of state has to be blocking)

        filtered_stream1 = stream1 -> filter(|value| {
            value <= context.state_ref(#sum_of_stream2_reference).borrow_mut().as_reveal_ref()
        }) -> for_each(|x| println!("{}", x));
    };

    df.run_available(); // Should return quickly and not hang
}
