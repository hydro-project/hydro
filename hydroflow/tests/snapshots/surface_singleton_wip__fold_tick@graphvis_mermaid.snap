---
source: hydroflow/tests/surface_singleton_wip.rs
expression: "df.meta_graph().unwrap().to_mermaid(&Default::default())"
---
%%{init:{'theme':'base','themeVariables':{'clusterBkg':'#ddd','clusterBorder':'#888'}}}%%
flowchart TD
classDef pullClass fill:#8af,stroke:#000,text-align:left,white-space:pre
classDef pushClass fill:#ff8,stroke:#000,text-align:left,white-space:pre
classDef otherClass fill:#fdc,stroke:#000,text-align:left,white-space:pre
linkStyle default stroke:#aaa
1v1[\"(1v1) <code>source_iter(10..=30)</code>"/]:::pullClass
2v1[\"(2v1) <code>source_iter_delta(15..=25)</code>"/]:::pullClass
3v1[\"(3v1) <code>map(Max::new)</code>"/]:::pullClass
4v1[\"(4v1) <code>lattice_reduce()</code>"/]:::pullClass
5v1[/"(5v1) <code>tee()</code>"\]:::pushClass
6v1[\"(6v1) <code>inspect(|x| println!(&quot;inspect {}&quot;, x))</code>"/]:::pullClass
7v1[\"(7v1) <code>identity::&lt;Max&lt;_&gt;&gt;()</code>"/]:::pullClass
8v1[\"(8v1) <code>zip()</code>"/]:::pullClass
9v1[\"(9v1) <code>filter(|(value, sum_of_stream2)| { value &lt;= sum_of_stream2.as_reveal_ref() })</code>"/]:::pullClass
10v1[/"(10v1) <code>for_each(|x| println!(&quot;filtered {:?}&quot;, x))</code>"\]:::pushClass
11v1[/"(11v1) <code>for_each(|x: Max&lt;_&gt;| println!(&quot;state {:?}&quot;, x))</code>"\]:::pushClass
12v1["(12v1) <code>handoff</code>"]:::otherClass
13v1["(13v1) <code>handoff</code>"]:::otherClass
2v1-.->3v1; linkStyle 0 stroke:#060
4v1==>5v1; linkStyle 1 stroke:#060
3v1-.->12v1; linkStyle 2 stroke:#060
6v1-->|0|8v1
1v1-->6v1
7v1-->|1|8v1
5v1==>13v1; linkStyle 6 stroke:#060
9v1-->10v1
8v1-->9v1
5v1==>11v1; linkStyle 9 stroke:#060
12v1-.->4v1; linkStyle 10 stroke:#060
13v1==>7v1; linkStyle 11 stroke:#060
subgraph sg_1v1 ["sg_1v1 stratum 0"]
    2v1
    3v1
    subgraph sg_1v1_var_stream2 ["var <tt>stream2</tt>"]
        2v1
        3v1
    end
end
subgraph sg_2v1 ["sg_2v1 stratum 0"]
    1v1
    6v1
    7v1
    8v1
    9v1
    10v1
    subgraph sg_2v1_var_filtered_stream1 ["var <tt>filtered_stream1</tt>"]
        6v1
    end
    subgraph sg_2v1_var_filtered_stream2 ["var <tt>filtered_stream2</tt>"]
        8v1
        9v1
        10v1
    end
    subgraph sg_2v1_var_stream1 ["var <tt>stream1</tt>"]
        1v1
    end
end
subgraph sg_3v1 ["sg_3v1 stratum 0"]
    4v1
    5v1
    11v1
    subgraph sg_3v1_var_sum_of_stream2 ["var <tt>sum_of_stream2</tt>"]
        4v1
        5v1
    end
end

